### ISSUE TITLE
[prefer-readonly-parameter-types] Proposal: Add type allowlist

### ISSUE THREAD

**Issue by marekdedic** [typescript-eslint/typescript-eslint#4185]
This propsal is a continuation of the discussion in #4061, cc @bradzacher 

The rule `prefer-readonly-parameter-types` is known to not play nicely with certain types, with the canonical example being the `HTMLElement` from the standard lib. This type cannot be made deep-readonly, as that breaks TS compilation altogether.

# Outstanding questions
- Should only the types themselves be ignored, or should the check ignore them, even if they are part of a more complex type?

**Comment by bradzacher** [comment]
Copying my comment here for additional context:

We can create an allowlist of types that are problematic and expand on them as required.
This should be pretty straightforward to implement.

1) get the `Symbol` for the type
2) get the declarations for the type
3) ensure that the declaration is within a TS "default library"
    - This check is to ensure that we don't false-negative if someone declares a `HTMLElement` type locally within their file

For example here is some code we could generalise from an existing rule:
https://github.com/typescript-eslint/typescript-eslint/blob/71c93706e55f5f92a1285102b93c6ab1950c6df4/packages/eslint-plugin/src/rules/no-implied-eval.ts#L83-L91

**Comment by marekdedic** [comment]
As far as the subtype goes, I would allow that as well - e.g.
```ts
const x = {a: readonly number; b: {c: readonly string; d: HTMLElement;}; }
```
would pass (the rule would only skip `x.b.d` and still check the other field).

**Comment by marekdedic** [comment]
Hi,
could this issue please be marked as accepting PRs? Or are there any changes needed? I may have some time over Christmas for FOSS contributions, maybe this could be one of them :)

**Comment by RebeccaStevens** [comment]
Just thought I'd document this somewhere (the important parts should probably be added to the docs).

[type-fest](https://www.npmjs.com/package/type-fest)'s [`ReadonlyDeep`](https://github.com/sindresorhus/type-fest/blob/main/source/readonly-deep.d.ts) type does its best to convert any given type to an immutable type but currently fails for objects that contain both properties and call signatures. 
I've submit a PR [type-fest#359](https://github.com/sindresorhus/type-fest/pull/359) to add support of a handling the case of single call signature with properties but it seems that due to a TypeScript limitation, multiple call signature cannot be supported. This issue is reported in [TypeScript#29732](https://github.com/microsoft/TypeScript/issues/29732).

`allowlist` will need to be used to for these type to prevent `isTypeReadonly` from returning false for them.
Note: `treatMethodsAsReadonly` isn't enough for this as `ReadonlyDeep<T>` just returns `T` for objects that have call signatures. Thus all properties will also stay mutable.

Another thing to note is that as of now, `treatMethodsAsReadonly` needs to be on when using `ReadonlyDeep` for maps and sets, at least until [type-fest#358](https://github.com/sindresorhus/type-fest/pull/358) is merged.

**Comment by marekdedic** [comment]
Fixed in #4436, due to be released in 6.0.
